### DISCUSSION TITLE
Writing a rule to cycle colours endlessly

### DISCUSSION THREAD

**Discussion by WilliWissler** [arendst/Tasmota#22132] [Support request]
I want to write a rule that will trigger on power on and fade from R0 G0 B255 to R255 G0 B0 and back continuously in a ndless loop. 

I want this to be saved and run whenever thr bulb is turned on.

I looked at the notes on rule scripting and got chat gpt yo write the below code, amongst variations. But it only will make one transition even when triggered manually in the console on tasmotrol, so not sure what the problem is.

Rule1
  ON System#Boot DO
    # Ensure the device starts with red color and fades
    IF (Power1#State=0) OR (Color1#State!=255,0,0) THEN
      Power1 ON; Color1 255,0,0; Fade 1; Speed 150; RuleTimer1 15
    ENDIF
  ENDON

  ON Rules#Timer=1 DO
    # Transition to blue after 15 seconds
    IF (Color1#State=255,0,0) THEN
      Color1 0,0,255; Fade 1; Speed 150; RuleTimer2 15
    ENDIF
  ENDON

  ON Rules#Timer=2 DO
    # Transition back to red after 15 seconds
    IF (Color1#State=0,0,255) THEN
      Color1 255,0,0; Fade 1; Speed 150; RuleTimer1 15
    ENDIF
  ENDON

DEVICE DETAILS
Tasmota flashed bulb (Esp8285) from athom (https://www.athom.tech/blank-1/color-bulb). It seems to have the tasmota.bin firmware. Accessing through tasmotrol on android.

**Comment by arendst** [comment]
You did play with the ``scheme`` command?

**Reply by WilliWissler** [reply]
Im not sure what you mean sorry. Im a total noob.

**Comment by arendst** [comment]
In that case start here https://tasmota.github.io/docs/

Look for command ``scheme``

**Reply by WilliWissler** [reply]
I couldn't find a way to go just blue to red with a scheme number. Am I missing something?

**Reply by sfromis** [reply]
You can set it up with the commands `Palette`, `Speed` and `Fade`, for your `Scheme` to have something to work with, without messing with any rules.

And your rules are broken when you do not follow the docs, but instead do it by wishful thinking. You cannot use those trigger references with `#` anywhere but in the trigger part of a rule. Also missing `Backlog`, and inventing an non-existing keyword `then`. Did you ask ChatGPT for bad advice?

**Comment by WilliWissler** [comment]
So I used just a command in the console:

Backlog Scheme 2; Palette FF0000 0000FF; Speed 40;

This returned 3x Command Unknown, and no change on the bulb

I also read that palette doesn't save on flash so would I need a rule to set it on power on? (System#Boot DO Palette xxxxx ENDON)

**Reply by sfromis** [reply]
Yes, The [docs](https://tasmota.github.io/docs/Commands/#palette) very clearly says that you can use such a rule to set the `Palette` again at boot.

Copy/pasting your command into the Tasmota console worked, both with a PWM-based RGB bulb and a WS2812-based strip.
`Backlog Scheme 2; Palette FF0000 0000FF; Speed 40;`
Since you say "3x Command Unknown", that sounds like you have no light defined, and your initial post strongly suggested that you were already using the `Speed` command, which is not at all consistent with it now not being accepted.

Depending on where you copy/pasted from (or what you did to create the command), your computer may be playing tricks messing up the text. If you enable `Weblog 3` before trying the command again, Tasmota will produce more detailed logs, including showing the command names processed.

**Reply by WilliWissler** [reply]
I must have removed some device setup stuff when I entered a "reset" command instead of restart, I reentered the configuration code from blackadder and it seems to have restored the Ui in tasmota and also allowed me to enter those commands. 

The bulb was then switching between blue and red but not fading. I turned on fade and it does kinda fade, but also flickers and jumps between colours (like from magenta to red, purple to blue and in other spots along the transition back and forth. 

Is this flickering and jumping something that can be smoothed out with different code or configuration?

**Reply by sfromis** [reply]
My experience testing your specific command sequence above, obviously also with `Fade 1`, was smooth slow fading, in both types of light I tested. Not perceiving any steps or flickering. Not having any rules or other command sources to intervene.

What lamp type is your config?

**Comment by WilliWissler** [comment]
Not sure exactly what you mean by config lamp type sorry. The bulb is an Athom bulb (https://www.athom.tech/blank-1/color-bulb) that foes warm cold white light as well as rgb, soes that make it RGBCCT. Is that what you mean? Or is this information I can find in the Tasmotrol app?

**Reply by sfromis** [reply]
Yes, 5-channel PWM (RGBCW), same type light as one of my test cases.
https://templates.blakadder.com/athom_LB017W.html

Should work fine with smooth fading (like I'm seeing), as long as there is not more getting in the way. Watching the console, after having activated `Weblog 3` for details, is generally a good place to see what happens.

**Comment by WilliWissler** [comment]
I managed to get it smoother somehow tonight but when restarting the palette was still lost, so rewrote the rule command as "system#init do Palette FF00000000 0000FF0000 endon" and now it seems to be working when reset the motion appears smoother but maybe just jumpy at the ends. Maybe this is because it is a triangular oscilation rather than more sine shaped. Is there any way with scheme to affect the pacing outside of speed?

**Reply by sfromis** [reply]
Maybe if you disabled gamma correction could make the steps coarse at low brightness.

Anyway, test case, more detailed than probably needed:
`Backlog Scheme 0; Palette 0; Speed 32; Fade 0; Dimmer 50; Power off; HsbColor`
Power cycle the device
`Backlog Palette FF0000 0000FF; Speed 40; Fade 1; Dimmer 50; Scheme 2; Power on; HsbColor`
Maybe repeat if it does not keep cycling.

**Reply by WilliWissler** [reply]
That works fine but still jumps a bit after reaching the maximum and minimum of the transition. Do you know what would be a more advanced way to gain control of the timing of transitions?

**Reply by WilliWissler** [reply]
Modifying the list of hex codes to include some intermediary colours has allowed me to get the motion a bit smoother. Just wondering how additional colours affects the timing. I guess I would be able  to create more complex timings be adding more intermediary steps. Would it be correct that the more coloura I add the quicker it cycles through each one and if i repeat a colour it will pause on that colour for whatever fraction of time that step is based on the number of total colours?
